src/forms/EditListingAvailabilityExceptionForm/EditListingAvailabilityExceptionForm.js

```diff
@@ -0,0 +1,670 @@
+import React, { useState } from 'react';
+import { array, bool, func, shape, string } from 'prop-types';
+import { compose } from 'redux';
+import { Form as FinalForm } from 'react-final-form';
+import classNames from 'classnames';
+
+import { intlShape, injectIntl, FormattedMessage } from '../../util/reactIntl';
+import { propTypes } from '../../util/types';
+import {
+  getSharpHours,
+  getStartHours,
+  getEndHours,
+  isDayMomentInsideRange,
+  isInRange,
+  isSameDate,
+  isSameDay,
+  resetToStartOfDay,
+  timeOfDayFromLocalToTimeZone,
+  timeOfDayFromTimeZoneToLocal,
+  dateIsAfter,
+  findNextBoundary,
+  timestampToDate,
+  localizeAndFormatTime,
+  monthIdStringInTimeZone,
+  getMonthStartInTimeZone,
+  nextMonthFn,
+  prevMonthFn,
+} from '../../util/dates';
+import { bookingDateRequired } from '../../util/validators';
+import {
+  FieldDateInput,
+  FieldRadioButton,
+  FieldSelect,
+  Form,
+  IconArrowHead,
+  PrimaryButton,
+} from '../../components';
+
+import css from './EditListingAvailabilityExceptionForm.css';
+
+const TODAY = new Date();
+const MAX_AVAILABILITY_EXCEPTIONS_RANGE = 365;
+
+// Date formatting used for placeholder texts:
+const dateFormattingOptions = { month: 'short', day: 'numeric', weekday: 'short' };
+
+// React-dates returns wrapped date objects
+const extractDateFromFieldDateInput = dateValue =>
+  dateValue && dateValue.date ? dateValue.date : null;
+
+// Get first available start time as timestamp or null.
+const firstAvailableStartTime = availableStartTimes =>
+  availableStartTimes.length > 0 && availableStartTimes[0] && availableStartTimes[0].timestamp
+    ? availableStartTimes[0].timestamp
+    : null;
+
+const endOfAvailabilityExceptionRange = (timeZone, date) => {
+  return resetToStartOfDay(date, timeZone, MAX_AVAILABILITY_EXCEPTIONS_RANGE - 1);
+};
+
+// Ensure that AvailabilityExceptions are in sensible order.
+// NOTE: this sorting can only be used for non-paginated list of exceptions (page size is max 100)
+const sortExceptionsByStartTime = (a, b) => {
+  return a.attributes.start.getTime() - b.attributes.start.getTime();
+};
+
+// Convert exceptions list to inverted array of time-ranges that are available for new exceptions.
+const getAvailableTimeRangesForExceptions = (exceptions, timeZone) => {
+  const nextBoundary = findNextBoundary(timeZone, TODAY);
+  const lastBoundary = endOfAvailabilityExceptionRange(timeZone, TODAY);
+
+  // If no exceptions, return full time range as free of exceptions.
+  if (!exceptions || exceptions.length < 1) {
+    return [{ start: nextBoundary, end: lastBoundary }];
+  }
+
+  const endOfLastException = exceptions[exceptions.length - 1].attributes.end;
```

Comment:
Should we get the last exception from the sortedExceptions list instead? In case the exceptions are not already in order.